exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -489,6 +492,10 @@ func getNewMachine(mp *expv1.MachinePool, infraMachine *unstructured.Unstructure
 		},
 	}
 
+	// This object is passed into ensureInfraMachineOnwerRefs, so we need to set the TypeMeta in this case.
+	// If we pulled the Machine from a List() or Get(), the TypeMeta would already be set.
+	machine.SetGroupVersionKind(clusterv1.GroupVersion.WithKind("Machine"))
```

Comment:
seems like there are a few changes outside of the Docker infra provider in this PR, can those be split out? Do they need to be backported?

Reply:
Sure, I can put that in a separate PR. I believe it shouldn't break anything but let me test that out.

Reply:
I'd like to still include `utilexp.GetMachinePoolByLabels()`, though, since it's used in dockermachine_controller.go.

Reply:
@Jont828 I agree with @CecileRobertMichon here, I have annotated 3 places where I think the PR is not about core logic of machinepool machine implementation, it goes into the actual docker machine pool implementation, which will make it difficult for provider implements to use this PR as a reference PR. I think your initial PR was much more aligned as a reference PR which I copied from, now this looks quite complicated that we actually implemented in OCI. I think we should consider splitting this to multiple PRs , first implement any missing logic in docker machinepool and then implement the machine pool machine on top of it as reference PR.

Reply:
Since this a wider impact comment, I think we should consider having a separate type for machine pool machine, I think in the PR, we use infrav1.DockerMachine, which I think is common with the non machinepool machine as well, which may get into problems later, for example what if we want to watch the machinepool machines and delete them from the underlying provider? This is just my thought though, this is how we implemented in OCI as a separate machinepool machine type.

Reply:
Thanks for the feedback @shyamradhakrishnan. We're reusing the DockerMachine type here for InfraMachinePoolMachines as Docker doesn't have an underlying MachinePool type, i.e. all MP replicas are just Docker containers. I would expect some providers to create their own type, like AzureMachinePoolMachine in CAPZ, and use that by setting the `InfrastructureMachineKind` to refer to a separate CRD if needed.

Reply:
The original implementation for DockerMachinePools had to reuse/copy a lot of the code from the DockerMachine controller for stuff like bootstrap and setting addresses, but now that we are creating DockerMachines, we can reuse the DockerMachine controller to do most of the heavy lifting. That way, we can drop a lot of duplicate code. As for the original PR, I also discovered a bug/error in how I was handling deletion, which also contributed to the need for this refactor. In the initial PR, I was deleting the Docker container and then trying to delete the Machine, which caused a failure as we were unable to drain the nodes. Instead, when we want to delete a Docker container, we want to delete the Machine to trigger node cordon/drain and InfraMachine deletion.

Reply:
Essentially, the Infra MPM flow I'm implementing is as follows:

- When we scale up, create the instances on the Infra provider, i.e. VMSS VM or Docker container.
- Create InfraMachines to represent each instance created by the Infra provider.
- When we scale down, select an instance for deletion and delete the instance's owner Machine, and handle the edge cases where the owner Machine doesn't exist.

I feel that the old NodePool struct in nodepool.go became increasingly awkward to integrate with the solution outlined above, as it dealt with containers directly and could not maintain its own inventory of information like the providerID or bootstrap status and had to rely on the DockerMachine type. Additionally, it would have been extremely awkward to try to handle the deletion flow of deleting clusterv1.Machines in nodepool.go as well.